packages/flutter/lib/src/material/reorderable_list.dart

```diff
@@ -0,0 +1,402 @@
+// Copyright 2018 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:math';
+
+import 'package:flutter/widgets.dart';
+import 'package:flutter/rendering.dart';
+
+import 'material.dart';
+
+/// A callback used by [ReorderableListView] to swap items in a list.
+/// 
+/// Implementations should remove the corresponding list item at [oldIndex]
+/// and insert a new one at [newIndex].
+/// 
+/// Note that if [oldIndex] is before [newIndex], removing the item at [oldIndex]
+/// from the list will reduce the list's length by one. Implementations used
+/// by [ReorderableListView] will need to account for this when inserting before
+/// [newIndex].
+/// 
+/// Example implementation:
+/// 
+/// ```dart
+/// final List<MyDataObject> backingList = <MyDataObject>[/* ... */];
+/// void onSwap(int oldIndex, int newIndex) {
+///   if (oldIndex < newIndex) {
+///     // removing the item at oldIndex will shorten the list by 1.
+///     newIndex -= 1;
+///   }
+///   final MyDataObject element = backingList.removeAt(oldIndex);
+///   backingList.insert(newIndex, element);
+/// }
+/// ```
+typedef void OnSwapCallback(int oldIndex, int newIndex);
+
+/// A list with draggable content that the user can re-order.
+/// 
+/// Note that this widget places its [children] in a [Column] or [Row] and not a [ListView].
+///
+/// All [children] must have a key.
+class ReorderableListView extends StatefulWidget {
+
+  /// Creates a reorderable list.
+  const ReorderableListView({
+    @required this.children, 
+    @required this.onSwap, 
+    this.scrollDirection = Axis.vertical, 
+    this.padding, 
+    this.dropAreaExtent = 96.0,
+  }): assert(dropAreaExtent != null),
+      assert(scrollDirection != null),
+      assert(onSwap != null),
+      assert(children != null);
+
+  /// The widgets to display.
+  final List<Widget> children;
+
+  /// The [Axis] along which the list scrolls.
+  /// 
+  /// List children also drag along this [Axis].
+  final Axis scrollDirection;
+
+  /// The amount of space by which to inset the child.
+  final EdgeInsets padding;
+
+  /// Called when a list child is dropped into a new position to shuffle the
+  /// underlying list.
+  /// 
+  /// This [ReorderableListView] calls [onSwap] after a list child is dropped
+  /// into a new position.
+  final OnSwapCallback onSwap;
+
+  /// The extent along the [scrollDirection] axis to allow a child to drop
+  /// into when the user reorders list children.
+  /// 
+  /// When the user is dragging a child to reorder it, the list will open up a
+  /// drop area in the space under the widget being dragged. That area will
+  /// have an extent along the main axis of [dropAreaExtent].
+  /// 
+  /// In a vertical list, this should match the height of the list items.
```

Comment:
Since the height of the list items can vary, I assume that you mean "match the typical list item height".  

Since the drag occurs after layout, we should the actual sizes of the list items.